### ISSUE TITLE
Leverage code generation

### ISSUE THREAD

**Issue by saschatimme** [JuliaAlgebra/FixedPolynomials.jl#9]
I just had the idea, that similar to the horner macro in Base it should actually be possible to
auto generate code for the efficient evaluation of multivariate polynomials.
For example it should be possible for the polynomial `x^3*y^2+4x^2+3` to generate the code
```julia
function _eval(xs)
   x, y = xs
   x_2 = x * x
   x_3 = x_2 * x
   y_2 = y * y

   x_3 * y_2 + 4 * x_2 + 3
end
```

Do you think this is something worth exploring @blegat ?

**Comment by blegat** [comment]
It could be interesting indeed, however it will be a function returning an expression and not a macro since the code generated depends on the value of the polynomial. In fact, it could be implemented as a generated function for @rdeits' [StaticPolynomials](https://github.com/rdeits/StaticPolynomials.jl) since the exponents are in the monomial type.

**Comment by saschatimme** [comment]
So if I understand correctly, the critical part is that somehow all monomials with their corresponding exponent need to be in the type itself right?

**Comment by blegat** [comment]
I wouldn't say that it is necessary, you could have a function that returns a function to evaluate the polynomial using this technique.
For StaticPolynomials, it seems you can do this directly on the `substitute` function.

**Comment by saschatimme** [comment]
I thought a little bit more about and considered a whole bunch of different approaches, but I think I will try the following:
```julia
function generate_evaluate(f)
     N = nterms(f)
     E = exponent_matrix_as_expression(f)
     ID = Val{Base.gensym()}

     x -> begin
          magic_evaluate(x, coefficients(f), E, Val{N}, ID)
     end
end

@inline @generated function magic_evaluate(x, coeffs, expr, ::Type{Val{N}}, ::Type{Val{ID}}) where {N, ID}
     # transform the exponents expr together with x in instructions for evaluation
     # of the term monomials (as an expression still)
     instr = evaluation_instructions_expr(x, expr)

     # stitch together the result using coeffs, and N (to know how many indices coeffs has)
     evaluate_everything(instr, coeffs, Val{N})
end
```
The `ID` is necessary since the generated function does not distinguish between different expressions.

Obviously, I don't have an idea whether this will really work, but I think I will give it a shot this evening :) (and probably find out that I overlooked something)

**Comment by blegat** [comment]
That looks promising, looking forward to see the results !

**Comment by saschatimme** [comment]
Obviously the approach above did not work (you can only access type informations) in the `@generated` function (and not the content of `expr`)...

**But** I came up with something else. Which is kinda crazy, but it seems to work.


I defined a type
```
struct SuperPolynomial{T, NVars, NTerms, Exponents<:Val}
    coefficients::Vector{T}
end
```
where the Exponents are stored as a Tuple (the exponents matrix is flattened) in a `Val` type.
E.g. `[1 0; 2 4]` is stored as `Val{(1, 2, 0, 4)}`.

With this information I can unroll everything, eliminate unnecessary multiplications etc.

The generated code for the example above would be
```julia
begin
    _x1 = x[1]
    _x1_1 = _x1
    _x2 = x[2]
    _x2_2 = pow(_x2, 2)
    _x2_4 = _x2_2 * pow(_x2, 2)
    sum = zero(Float64)
    @inbounds sum = muladd(f.coefficients[1], _x2_4, sum)
    @inbounds sum = muladd(f.coefficients[2] * _x1_1, _x2_2, sum)
    sum
end
```
and the compiler can optimize like crazy :o 

Here is an example with timings

```julia
import SuperPolynomials
const SP = SuperPolynomials
import FixedPolynomials
const FP = FixedPolynomials
using BenchmarkTools

# We have 24 variables
N = 24
# with 20 terms
M = 20
# generate random exponents with degree between 0 and 5 (per variable)
exponents = round.(Int, 5 * rand(N, M))
# some random coefficients
coefficients = rand(M)
g = SP.SuperPolynomial(coefficients, exponents)
w = rand(N)

f = FP.Polynomial(exponents, coefficients)
cfg = FP.GradientConfig(f)
# @btime SP.horner($g, $w)
@btime SP.evaluate($g, $w) # 130.837 ns (0 allocations: 0 bytes)
@btime FP.evaluate($f, $w, $cfg) # 1.580 μs (0 allocations: 0 bytes)
```
That's a factor of **12**.

I also implemented a multivariate Horner scheme given by some paper, but somehow the generated
functions allocates if the exponents matrix is too large (and made me questions everything I did^^). I really do not understand why... 

You can find the code at [SuperPolynomials.jl](https://github.com/saschatimme/SuperPolynomials.jl).
The strange Horner function is also there, maybe you have an idea why it behaves differently.

This stuff is obviously completely type instable. To support systems of polynomials in this way
I will probably end up with a design which leverages `FunctionWrappers.jl` and closures...

**Comment by blegat** [comment]
Wow that's impressive, it really shows the power of Julia. I don't have a lot of time to investigate this right now but I am glad to see that you have got such impressive speedup.

**Comment by saschatimme** [comment]
Things get usable. The allocations mentioned above are actually a Julia bug ([#14821](https://github.com/JuliaLang/julia/issues/14821)).

I also implemented the evaluation of the gradient, as well as an wrapper for systems of polynomials (up to 128 currently).
And of course conversions from MultivariatePolynomials and FixedPolynomials to SuperPolynomials (I need probably a proper name^^).

**Comment by cortner** [comment]
I've played with generating code for polynomials as well, and it worked very well for me. I gave up for entirely unrelated reasons, but am now getting back to it. `FunctionWrappers` was crucial for me as well. I also think `FixedPolynomials` really needs this, I was thinking of using it but there performance compared to generated code is very poor (factor 10-20 for most cases I've tested) as you've seen yourselves, so I'd be very happy to see a package that does this *well*.

**Comment by cortner** [comment]
I ran into problems when the polynomials became too big, i.e. too many multiplications and powers. Then suddenly Julia started allocating and I never understood why.

**Comment by rdeits** [comment]
There are internal limits on the maximum size of a tuple before the compiler gives up and stops specializing on its length. Since function arguments are internally tuples, that means there's also a limit on the number of function arguments that can be handled with no allocation. Also, I think `*` and `+` are special in that `a * b * c` lowers to `*(a, b, c)` instead of `*(*(a,b), c)`, which could potentially trigger this upper limit for complex polynomials. So perhaps you could manually do that grouping into binary ops instead of asking Julia to multiply all of the arguments at once?

**Comment by cortner** [comment]
thank you - that is very helpful to know. (I guess this is the lay-man's explanation of the Julialang GitHub issues sombody - maybe you - pointed me to).

@dussong  this is our problem. We basically need to rewrite our monster-polynomials with lots of brackets, or we can try using `muladd` and friends.

**Comment by cortner** [comment]
Has this became `StaticPolynomials`? I just tried it on a first simple use-case and it looks fast. I particularly love the fact that you have efficient Jacobians. I'll keep experimenting with this.

**Comment by saschatimme** [comment]
Hey @cortner,

yes `StaticPolynomials` is the result of this idea and I am very happy with the current results. I use this already in the current development version of [HomotopyContinuation.jl](https://github.com/JuliaHomotopyContinuation/HomotopyContinuation.jl/tree/dev). Obviously you have the compiler overhead (which seems to get much better in Julia 0.7), but for the right application I think this is the way to go. At some point I wrote an (internal) announcement about StaticPolynomials which contains some Benchmarks. It is available [here](https://gist.github.com/saschatimme/4c089f8d1454aac1cf24a9b811e26050). An additional benefit of this approach is that I used a different algorithm than for FixedPolynomials. This algorithm should have a much better numerically accuracy and additionally I could use quite a lot of FMA instructions which additionally should improve things.

If you have any questions just let me know.
